### DISCUSSION TITLE
Connecting OutlineViewerInput to Language Server After Restart

### DISCUSSION THREAD

**Discussion by sdaimwood** [eclipse/lsp4e#633] [Q&A] [answered]
I have a question I’m hoping the community can answer, I’m attempting to implement functionality for our editor where if certain user settings are changed, we stop the language server and allow it to restart to pick up the new settings. So far, I’ve got everything working after the server starts back up, except the outline view which fails with a `SocketException: Socket closed` error (see screenshot at bottom).
 
Closing the document all together and reopening, or closing and reopening the outline view fixes the issue, but would prefer to handle it for users if possible.
 
For reference, we’re using version 0.20.7 due to java 11 requirements.
 
Overview of steps that I’m taking:
- Get LS Wrapper for editor
- Call wrapper.stop()
- LS exits, sockets are closed
- Document tab is reopened from settings, server startup command is sent
- Functionality is working, minus the outline view which shows the `SocketException` error
 
The error is cropping up in the `LSSymbolsContentProvider` class at line 308 (for 0.20.7), seemingly because the `languageServer` object in `outlineViewerInput` is never updated after the original language server is closed. I’ve searched around for ways to update `outlineViewerInput` to point to the new LS programmatically but haven’t been able to figure it out. Any ideas on how to solve this issue or is this sort of function not possible?
 
Screenshot:
 
![image](https://github.com/eclipse/lsp4e/assets/93732205/0b9b4ee1-0e9f-4ff9-838d-488ec193082d)

**Comment by angelozerr** [comment, marked as answer]
> I have a question I’m hoping the community can answer, I’m attempting to implement functionality for our editor where if certain user settings are changed, we stop the language server and allow it to restart to pick up the new settings.

To avoid stoping the language server, why cannot use workspace/didChangeConfiguration?

**Reply by sdaimwood** [reply]
> To avoid stoping the language server, why cannot use workspace/didChangeConfiguration?

I hadn't thought about that. The settings themselves are on the client side and aren't stored in the language server at all, but they do impact functionality. The primary purpose of closing down the server is to clear the cache, but if a full restart is not possible, I could try using a `didChangeConfiguration` call to clear the cache and reprocess the document. Thanks for the response, I'll give that a try.

**Reply by mickaelistria** [reply]
You may also want to include in your LS a LSP command to reset cache (the simplest would be via the `workspace/executeCommand`), and then add make Eclipse invoke this LS command when it seems necessary.

**Reply by sdaimwood** [reply]
Thank you for the suggestions. I ended up actually implementing both, the `workspace/didChangeConfiguration` to clear the cache for the whole LS, and then a `workspace/executeCommand` to tell the server to re-parse open documents (which picks up the new settings). Thanks again.

**Reply by angelozerr** [reply]
I wonder why you need to implement a workspace/executeCommand? Why you cannot react to the settings change in workspace/didChangeConfiguration?

You can check if your settings change and in this case, you can update the cache and publish diagnostics again. It will avoid to implement a custom client command and will work with any LSP clients.
